File: src/main/java/com/snackgame/server/common/exception/dto/ExceptionResponse.java
package com.snackgame.server.common.exception.dto;

import java.util.List;

public class ExceptionResponse {

    private static final String DEFAULT_MESSAGE = "처리 중 예외가 발생했습니다";

    private final String action;
    private final List<String> messages;

    public ExceptionResponse(String message) {
        this.action = null;
        this.messages = List.of(message);
    }

    public ExceptionResponse(List<String> messages) {
        this.action = null;
        this.messages = messages;
    }

    public ExceptionResponse(String action, String message) {
        this.action = action;
        this.messages = List.of(message);
    }

    public static ExceptionResponse withDefaultMessage() {
        return new ExceptionResponse(DEFAULT_MESSAGE);
    }

    public static ExceptionResponse from(Exception exception) {
        return new ExceptionResponse(exception.getMessage());
    }

    public String getAction() {
        return action;
    }

    public List<String> getMessages() {
        return messages;
    }
}

File: src/main/java/com/snackgame/server/common/exception/dto/ExceptionResponse.kt
package com.snackgame.server.common.exception.dto

import org.springframework.http.HttpStatus

data class ExceptionResponse(
    val messages: List<String> = DEFAULT_MESSAGE,
    val code: String = DEFAULT_CODE,
) {
    constructor(messages: List<String>, throwableClazz: Class<Throwable>) : this(
        messages,
        throwableClazz.simpleName.upperSnakeCase()
    )

    constructor(message: String, throwableClazz: Class<Throwable>) : this(
        listOf(message),
        throwableClazz
    )

    companion object {
        private val DEFAULT_MESSAGE = listOf("처리 중 예외가 발생했습니다")
        private val DEFAULT_CODE = HttpStatus.INTERNAL_SERVER_ERROR.name
    }
}

private fun String.upperSnakeCase(): String {
    return buildString {
        for (char in this@upperSnakeCase) {
            if (char.isUpperCase() && this.isNotEmpty()) {
                append('_')
            }
            append(char.uppercaseChar())
        }
    }
}
